packages/ecmascript/src/Shims.ts

```diff
@@ -124,30 +124,32 @@ export default class Shims {
 			case "Map":
 				arrayLikeClone = (arrayLike as Map<any, any>).entries();
 				break;
-			default: return arrayLikeClone;
-		}
-
-		if (mapfn) {
-			array = arrayLikeClone.map(mapfn);
-		} else {
-			array = arrayLikeClone;
+			case "Object":
+				arrayLikeClone = [];
+				// check whether the object is an array (using the length property)
+				if (Object.keys(arrayLike).find(item => item.indexOf('length') >= 0)) {
+					arrayLikeClone = Array.apply(null, Array(arrayLike));// nosonar
+				}
```

Comment:
I don't mind a rewritten version, but this does not do what I suggested!
```suggestion
				if (Object.hasOwn(arrayLike, 'length')) {
					arrayLikeClone = Array.apply(null, Array(arrayLike.length))
				}
```

Running both functions gives:
```
arrayFromPR({ length: 5 }, (v, i) => i); // Output: [0]
arrayFromPR({ length: 5 }) // Output: [{"length": 5}]
arrayFromMySuggestion({ length: 5 }, (v, i) => i); // Output: [0,1,2,3,4]
arrayFromMySuggestion({ length: 5 }) // Output: [undefined,undefined,undefined,undefined,undefined]
```

It does make it a bit more "compliant" to its type signature in that it now gives an array of the object. But it does not behave like the actual implementation.

Reply:
Re-reading the snippet, it seems you just missed the `.length`! :)
`arrayLikeClone = Array.apply(null, Array(arrayLike.length))`
`arrayLikeClone = Array.apply(null, Array(arrayLike));// nosonar`

Reply:
Thank you for your suggestion, however this does not compile with the older version of the `tslib`, hence the code is written in this way.

Reply:
Ah alright. I understand that the `Object.hasOwn` might not work, but without calling `Array.apply(null, Array(arrayLike.length))` this does not behave correctly. You called `Array(arrayLike)` not `Array(arrayLike.length)`

Reply:
Hi actually if you take a look at the screenshot it is behaving correctly with various types of objects. The `length` parameter is actually not required.

Reply:
Hi @Indy-rbo we discussed the changes internally, there is some difference in the behaviour, we will check it.

Reply:
Leaving an example to make things easier to understand:
```javascript
console.log(Array.from({ 0: 132, 2: "asd", length: 5 }));
// Logs: "[132, undefined, "asd", undefined, undefined]"
```

Reply:
As I was writing this out I saw the updates. It also includes a suggested change, so to hopefully assist I'll still post it:

---

Hi, I'd like to explain that this currently does not do what is expected.

Example 1 shows that `VROES.Shims.arrayFrom({ length: 5 });` gives `[{ length: 5 }]`. But  using the actual function it gives `[undefined, undefined, undefined, undefined, undefined]`

![image](https://github.com/user-attachments/assets/29abacc5-523c-4cfa-a06c-d1631865e703)

Example 2 is not valid, since you are actually passing in an Array which contains an object (and a number and a string). So that example does not hit the `Object` case.


The comment indicates there might be a misunderstanding of what my snippet did. 

`// check whether the object is an array (using the length property)`
We are not checking the length to check if the object is somehow also an array. We check the length since it's a special property for array-like objects. As the MDN docs say "array-like objects (objects with a length property and indexed elements)"

Reading the MDN documentation might help getting a clearer understanding of what it should do https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from

---

After reading it further myself, it turns out my suggestion was naive as the following case won't be handled:

```ts
Array.from({
    1: 'foo',
    2: 'bar',
    'non-indexed': 'baz',
    length: 5
})
```
Which should return:
`[undefined, 'foo', 'bar', undefined, undefined]`
![image](https://github.com/user-attachments/assets/bf4e6b55-4e77-4b8d-b500-f2fc4af83aa6)


So my adjusted snippet would be:

```ts
case "Object":
	arrayLikeClone = [];
	// Check if the object is an array-like object
	if (Object.keys(arrayLike).find(item => item.indexOf('length') >= 0)) {
		arrayLikeClone = Array.apply(null, Array(arrayLike.length));// nosonar
		Object.keys(arrayLike).forEach(element => {
			const indexKey = parseInt(element);
			if (!isNaN(indexKey) && indexKey < arrayLike.length) { // Check if object key is like an indexed element and within range
				arrayLikeClone[element] = arrayLike[element];
			}
		});
	}
	break;
```

Which in simple testing already came closer to the expected result:

![image](https://github.com/user-attachments/assets/b7d98089-33a7-4983-8714-cc8539e16fea)

Also if the index key is out of bounds it behaves as expected:
![image](https://github.com/user-attachments/assets/c8a5d3d2-7592-448a-9511-3d147397b789)

Reply:
Probably the easiest workaround here would be to create a new Array with the given length and after that add the objects for the given index:
```typescript
		const arrayLength = (arrayLike as any).length;
		arrayLikeClone = new Array(arrayLength);
		for (let i = 0; i < arrayLength; i++) {
			array[i] = arrayLikeClone[i];
		}
```
Something of this kind should do the magic.

@akantchev @Indy-rbo what do you think?

Reply:
@Indy-rbo I just saw your message above, yes, this is how I understand it as well. I think your code sample is better as it has more safety checks and should do the job, thank you!
@akantchev could you check the snippet from the reply just above mine? (https://github.com/vmware/build-tools-for-vmware-aria/pull/362#discussion_r1698241172)

Reply:
Hi @Indy-rbo , @VenelinBakalov.
After short internal discussion we figured out the exact issue and what code update is needed, code is updated to have exactly the result you mentioned.

Reply:
@Indy-rbo I think we Aleks' latest changes we should be good. Whenever you have a chance, please take a look as well
@akantchev could you just check one minor comment on the other changed file